core/sessions/oidcauth/oidc.go

```diff
@@ -0,0 +1,675 @@
+	}
+
+	var foundUser clsessions.User
+	err := sqlutil.TransactDataSource(ctx, oi.ds, nil, func(tx sqlutil.DataSource) error {
+		// Query the oidc user API token table for given token, user role and email are cached so
+		// no further upstream OIDC query is performed, sessions and tokens are synced against the upstream server
+		// via the UpstreamSyncInterval config and reaper.go sync implementation
+		var foundUserToken struct {
+			UserEmail string
+			UserRole  clsessions.UserRole
+			Valid     bool
+		}
+		if err := tx.GetContext(ctx, &foundUserToken,
+			"SELECT user_email, user_role, created_at + $2 >= now() as valid FROM oidc_user_api_tokens WHERE token_key = $1",
+			apiToken, oi.config.UserAPITokenDuration().Duration(),
+		); err != nil {
+			return err
+		}
+		if !foundUserToken.Valid {
+			return clsessions.ErrUserSessionExpired
+		}
+		foundUser = clsessions.User{
+			Email: foundUserToken.UserEmail,
+			Role:  foundUserToken.UserRole,
+		}
+		return nil
+	})
+	if err != nil {
+		if errors.Is(err, clsessions.ErrUserSessionExpired) {
+			// API Token expired, purge
+			if _, execErr := oi.ds.ExecContext(ctx, "DELETE FROM oidc_user_api_tokens WHERE token_key = $1", apiToken); execErr != nil {
+				oi.lggr.Errorf("error purging stale oidc API token session: %v", execErr)
+			}
+		}
+		return clsessions.User{}, err
+	}
+	return foundUser, nil
+}
+
+// ListUsers in the context of the OIDC driver only supports listing the local (admin) users, we don't have an identity server to query against
+func (oi *oidcAuthenticator) ListUsers(ctx context.Context) ([]clsessions.User, error) {
+	returnUsers := []clsessions.User{}
+	if err := sqlutil.TransactDataSource(ctx, oi.ds, nil, func(tx sqlutil.DataSource) error {
+		sql := "SELECT * FROM users ORDER BY email ASC;"
+		return tx.SelectContext(ctx, &returnUsers, sql)
+	}); err != nil {
+		oi.lggr.Errorf("error listing local users: %v", err)
+	}
+	return returnUsers, nil
+}
+
+// AuthorizedUserWithSession will return the API user associated with the Session ID if it
+// exists and hasn't expired
+func (oi *oidcAuthenticator) AuthorizedUserWithSession(ctx context.Context, sessionID string) (clsessions.User, error) {
+	if len(sessionID) == 0 {
+		return clsessions.User{}, errors.New("session ID cannot be empty")
+	}
+	var foundUser clsessions.User
+	err := sqlutil.TransactDataSource(ctx, oi.ds, nil, func(tx sqlutil.DataSource) error {
+		// Query the oidc_sessions table for given session ID, user role and email are saved after the id claims is provided and validated
+		var foundSession struct {
+			UserEmail string
+			UserRole  clsessions.UserRole
+			Valid     bool
+		}
+		if err := tx.GetContext(ctx, &foundSession,
+			"SELECT user_email, user_role, created_at + $2 >= now() as valid FROM oidc_sessions WHERE id = $1",
+			sessionID, oi.config.SessionTimeout().Duration(),
+		); err != nil {
+			return clsessions.ErrUserSessionExpired
```

Comment:
Should this confirm `sql.ErrNoRows` or something to rule out other kinds of errors?

Reply:
Yes we should catch that case, updated now to propagate up the error if its not the expected NoRows